test/common/src/main/kotlin/org/codice/compliance/utils/decorators/IdpResponseDecorators.kt

```diff
@@ -0,0 +1,72 @@
+/**
+ * Copyright (c) Codice Foundation
+ *
+ * <p>This is free software: you can redistribute it and/or modify it under the terms of the GNU
+ * Lesser General Public License as published by the Free Software Foundation, either version 3 of
+ * the License, or any later version.
+ *
+ * <p>This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details. A copy of the GNU Lesser General Public
+ * License is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+
+
+package org.codice.compliance.utils.decorators
+
+import org.codice.compliance.saml.plugin.IdpPostResponse
+import org.codice.compliance.saml.plugin.IdpRedirectResponse
+import org.codice.compliance.verification.binding.PostBindingVerifier
+import org.codice.compliance.verification.binding.RedirectBindingVerifier
+import org.w3c.dom.Node
+import javax.xml.parsers.DocumentBuilderFactory
+
+private fun buildDom(decodedSamlResponse: String?): Node? {
+    return DocumentBuilderFactory.newInstance().apply {
+        isNamespaceAware = true
+    }.newDocumentBuilder()
+            .parse(decodedSamlResponse?.byteInputStream())
+            .documentElement
+}
+
+/*
+ * These functions are responsible for returning the correct instance of
+ * a decorator from a response and
+ * a binding verifier from a decorator
+ */
+fun IdpRedirectResponse.decorate(): IdpRedirectResponseDecorator {
+    return IdpRedirectResponseDecorator(this)
+}
+
+fun IdpPostResponse.decorate(): IdpPostResponseDecorator {
+    return IdpPostResponseDecorator(this)
+}
+
+fun IdpRedirectResponseDecorator.bindingVerifier(): RedirectBindingVerifier {
+    return RedirectBindingVerifier(this)
+}
+
+fun IdpPostResponseDecorator.bindingVerifier(): PostBindingVerifier {
+    return PostBindingVerifier(this)
+}
+
+class IdpRedirectResponseDecorator(response: IdpRedirectResponse) : IdpRedirectResponse(response) {
```

Comment:
❗️ Because of the way this class and its sibling are structured, you have a lot of places in code where you _should_ be testing for null values for `responseDom`. Instead of the necessary tests, you've opted to use the `!!` operator which could result in NPEs. Avoiding null-class exceptions are one of the primary motivators for using Kotlin, so this is an unfortunate choice.

Recommend:
1. Change this class and its sibling to use late initialization and null protection for the `responseDom`
2. Remove all the places you've removed null safety with the `!!`.

Here's an example of how this decorator class could be altered:
```kotlin
class IdpRedirectResponseDecorator(response: IdpRedirectResponse) : IdpRedirectResponse(response) {
    var isRelayStateGiven: Boolean = false
    lateinit var decodedSamlResponse: String
    val responseDom: Node by lazy {
        checkNotNull(decodedSamlResponse)
        buildDom(decodedSamlResponse)
    }
}
```

Reply:
I REALLY like that `lazy` delegator.  When I was trying to use `lateinit` in some scenarios I was getting weird compilation errors, and laziness seems like a good substitute in those situations.

I should have used `lateinit` here though. Great catch. I'll go through and get rid of all of the !! if I can/it makes sense to do so!